File: libs/py/config/src/astra_config/kdl.py
"""Thin KDL → Python adapter (ckdl) shared by config + ontology loaders.

Keeps KDL strictly at the edge (CLAUDE.md): parse here, hand back plain dicts/scalars
+ `SecretRef`, and never thread raw ckdl nodes into application code. The ontology
loader (`libs/py/ontology`) reuses these helpers for its repeated-node walk.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

import ckdl

from .secrets import SecretRef

# ckdl ships no type stubs / py.typed, so its members (`parse`, `Document`, `Node`)
# aren't statically visible. Route access through an Any handle + alias the types;
# the runtime objects are real (see the probe in the Phase-1 spec).
_ckdl: Any = ckdl
KdlDocument = Any
KdlNode = Any


def load_document(path: str | Path) -> KdlDocument:
    """Parse a KDL file into a ckdl Document."""
    return _ckdl.parse(Path(path).read_text(encoding="utf-8"))


def node_props(node: KdlNode) -> dict[str, Any]:
    """A node's `key=value` properties as a plain dict."""
    return dict(node.properties)


def snake(name: str) -> str:
    """KDL kebab-case node name → Python snake_case key."""
    return name.replace("-", "_")


def leaf_value(node: KdlNode, *, secrets_file: Path | None = None) -> Any:
    """Resolve a leaf KDL node to a scalar, a `SecretRef`, a list, or None.

    * `name ref="sops:KEY"` → `SecretRef`
    * `name "x"` / `name 1` / `name #true` → the single typed arg
    * `name "a" "b"` → the list of args
    * `name` (no args) → None
    """
    props = node_props(node)
    if "ref" in props:
        return SecretRef(str(props["ref"]), secrets_file=secrets_file)
    args = list(node.args)
    if len(args) == 1:
        return args[0]
    if not args:
        return None
    return args


def children_as_dict(node: KdlNode, *, secrets_file: Path | None = None) -> dict[str, Any]:
    """A node's child nodes as `{snake_name: leaf_value | nested dict}`."""
    out: dict[str, Any] = {}
    for child in node.children:
        key = snake(child.name)
        if list(child.children) and not list(child.args) and "ref" not in node_props(child):
            out[key] = children_as_dict(child, secrets_file=secrets_file)
        else:
            out[key] = leaf_value(child, secrets_file=secrets_file)
    return out


def top_level_namespaces(
    doc: KdlDocument, *, secrets_file: Path | None = None
) -> dict[str, dict[str, Any]]:
    """Each top-level node → its children dict (the config-namespace shape)."""
    return {
        snake(node.name): children_as_dict(node, secrets_file=secrets_file) for node in doc.nodes
    }

File: libs/ts/config/src/kdl.ts
/**
 * Thin KDL → JS adapter (`@bgotink/kdl`) — the TS twin of `libs/py/config/kdl.py`.
 * Keeps KDL at the edge: parse here, return plain objects/scalars + `SecretRef`,
 * never thread raw KDL nodes onward. The ontology accessor reuses these helpers.
 */

import { readFileSync } from "node:fs";

import { type Document, type Node, parse } from "@bgotink/kdl";

import { SecretRef } from "./secrets";

export type LeafValue = string | number | boolean | bigint | null | SecretRef | unknown[];

export function loadDocument(path: string): Document {
  return parse(readFileSync(path, "utf8"));
}

/** A node's `key=value` properties as a plain object. */
export function nodeProps(node: Node): Record<string, unknown> {
  const props: Record<string, unknown> = {};
  for (const [key, entry] of node.getPropertyEntryMap()) props[key] = entry.getValue();
  return props;
}

/** KDL kebab-case node name → JS camelCase key. */
export function camel(name: string): string {
  return name.replace(/-([a-z0-9])/g, (_m, c: string) => c.toUpperCase());
}

function args(node: Node): unknown[] {
  return node.getArgumentEntries().map((e) => e.getValue());
}

function childNodes(node: Node): Node[] {
  return node.children?.nodes ?? [];
}

/** Resolve a leaf KDL node to a scalar, a `SecretRef`, an array, or null. */
export function leafValue(node: Node, secretsFile?: string): LeafValue {
  const props = nodeProps(node);
  if ("ref" in props) return new SecretRef(String(props.ref), secretsFile);
  const a = args(node);
  if (a.length === 1) return a[0] as LeafValue;
  if (a.length === 0) return null;
  return a;
}

/** A node's children as `{ camelName: leafValue | nested object }`. */
export function childrenAsObject(node: Node, secretsFile?: string): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const child of childNodes(node)) {
    const key = camel(child.name.name);
    const hasChildren = childNodes(child).length > 0;
    const hasArgs = child.getArgumentEntries().length > 0;
    if (hasChildren && !hasArgs && !("ref" in nodeProps(child))) {
      out[key] = childrenAsObject(child, secretsFile);
    } else {
      out[key] = leafValue(child, secretsFile);
    }
  }
  return out;
}

/** Each top-level node → its children object (the config-namespace shape). */
export function topLevelNamespaces(
  doc: Document,
  secretsFile?: string,
): Record<string, Record<string, unknown>> {
  const out: Record<string, Record<string, unknown>> = {};
  for (const node of doc.nodes) out[camel(node.name.name)] = childrenAsObject(node, secretsFile);
  return out;
}
